src/main/java/io/vertx/core/eventbus/impl/clustered/ClusteredEventBus.java

```diff
@@ -111,63 +104,50 @@ static void setTrustOptions(TCPSSLOptions sslOptions, TrustOptions options) {
   }
 
   @Override
-  public void start(Handler<AsyncResult<Void>> resultHandler) {
-    // Get the HA manager, it has been constructed but it's not yet initialized
-    HAManager haManager = vertx.haManager();
-    setClusterViewChangedHandler(haManager);
-    clusterManager.<String, ClusterNodeInfo>getAsyncMultiMap(SUBS_MAP_NAME, ar1 -> {
-      if (ar1.succeeded()) {
-        subs = ar1.result();
-        server = vertx.createNetServer(getServerOptions());
-
-        server.connectHandler(getServerHandler());
-        server.listen(asyncResult -> {
-          if (asyncResult.succeeded()) {
-            int serverPort = getClusterPublicPort(options, server.actualPort());
-            String serverHost = getClusterPublicHost(options);
-            serverID = new ServerID(serverPort, serverHost);
-            nodeInfo = new ClusterNodeInfo(clusterManager.getNodeID(), serverID);
-            vertx.executeBlocking(fut -> {
-              haManager.addDataToAHAInfo(SERVER_ID_HA_KEY, new JsonObject().put("host", serverID.host).put("port", serverID.port));
-              fut.complete();
-            }, false, ar2 -> {
-              if (ar2.succeeded()) {
-                started = true;
-                resultHandler.handle(Future.succeededFuture());
-              } else {
-                resultHandler.handle(Future.failedFuture(ar2.cause()));
-              }
-            });
+  public void start(Promise<Void> promise) {
+    server = vertx.createNetServer(getServerOptions());
+    server.connectHandler(getServerHandler());
+    server.listen(s -> {
+      if (s.succeeded()) {
+        int serverPort = getClusterPublicPort(options, server.actualPort());
+        String serverHost = getClusterPublicHost(options);
+        nodeInfo = new NodeInfo(serverHost, serverPort, options.getNodeMetadata());
+        nodeId = clusterManager.getNodeId();
+        Promise<Void> setPromise = Promise.promise();
+        setPromise.future().onComplete(ar -> {
+          if (ar.succeeded()) {
+            started = true;
+            deliveryStrategy.init(vertx);
+            promise.complete();
           } else {
-            resultHandler.handle(Future.failedFuture(asyncResult.cause()));
+            promise.fail(ar.cause());
           }
         });
+        clusterManager.setNodeInfo(nodeInfo, setPromise);
       } else {
-        resultHandler.handle(Future.failedFuture(ar1.cause()));
+        promise.fail(s.cause());
       }
     });
   }
 
   @Override
-  public void close(Handler<AsyncResult<Void>> completionHandler) {
-    super.close(ar1 -> {
+  public void close(Promise<Void> promise) {
+    Promise<Void> parentClose = Promise.promise();
+    super.close(parentClose);
+    parentClose.future().onComplete(ar -> {
```

Comment:
I think here we can benefit from flatmap for better readability

Reply:
Not sure how it would help here since we want to perform the server close whether or not the parent closed successfully.